packages/@aws-cdk/aws-ec2/lib/volume.ts

```diff
@@ -210,4 +215,470 @@ export enum EbsDeviceVolumeType {
+   * See {@link https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ebs-volume-types.html#EBSVolumeTypes_piops|Provisioned IOPS SSD (io1) volumes}
+   * for more information.
+   *
+   * This parameter is valid only for PROVISIONED_IOPS_SSD volumes.
+   *
+   * @default None -- Required for {@link EbsDeviceVolumeType.PROVISIONED_IOPS_SSD}
+   */
+  readonly iops?: number;
+}
+
+/**
+ * Attributes required to import an existing EBS Volume into the Stack.
+ */
+export interface VolumeAttributes {
+  /**
+   * The EBS Volume's ID
+   */
+  readonly volumeId: string;
+
+  /**
+   * The availability zone that the EBS Volume is contained within (ex: us-west-2a)
+   */
+  readonly availabilityZone: string;
+
+  /**
+   * The customer-managed encryption key that is used to encrypt the Volume.
+   *
+   * @default None -- The EBS Volume is not using a customer-managed KMS key for encryption.
+   */
+  readonly encryptionKey?: IKey;
+}
+
+/**
+ * Common behavior of Volumes. Users should not use this class directly, and instead use ``Volume``.
+ */
+abstract class VolumeBase extends Resource implements IVolume {
+  public abstract readonly volumeId: string;
+  public abstract readonly availabilityZone: string;
+  public abstract readonly encryptionKey?: IKey;
+
+  public grantAttachVolume(grantee: IGrantable, instances?: IInstance[]): Grant {
+    const result = Grant.addToPrincipal({
+      grantee,
+      actions: [ 'ec2:AttachVolume' ],
+      resourceArns : this.collectGrantResourceArns(instances),
+    });
+
+    if (this.encryptionKey) {
+      // When attaching a volume, the EC2 Service will need to grant to itself permission
+      // to be able to decrypt the encryption key. We restrict the CreateGrant for principle
+      // of least privilege, in accordance with best practices.
+      // See: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/EBSEncryption.html#ebs-encryption-permissions
+      const kmsGrant: Grant = this.encryptionKey.grant(grantee, 'kms:CreateGrant');
+      kmsGrant.principalStatement!.addConditions(
+        {
+          Bool: { 'kms:GrantIsForAWSResource': true },
+          StringEquals: {
+            'kms:ViaService': `ec2.${Stack.of(this).region}.amazonaws.com`,
+            'kms:GrantConstraintType': 'EncryptionContextSubset',
+          },
+        },
+      );
+    }
+
+    return result;
+  }
+
+  public grantAttachVolumeToSelf(instance: Instance, tagKeySuffix?: string): Grant {
+    const tagKey = `VolumeGrantAttach-${tagKeySuffix ?? this.stringHash(this.node.uniqueId)}`;
+    const tagValue = this.node.uniqueId;
+    const grantCondition: { [key: string]: string } = {};
+    grantCondition[`ec2:ResourceTag/${tagKey}`] = tagValue;
+
+    const result = this.grantAttachVolume(instance);
+    result.principalStatement!.addCondition(
+      'ForAnyValue:StringEquals', grantCondition,
+    );
+
+    Tag.add(this, tagKey, tagValue);
+    Tag.add(instance, tagKey, tagValue);
```

Comment:
Can you explain in a comment why the tag needs to be applied on `this`? That one surpised me.

For the other, don't forget to set ` { appliesToLaunchedInstances: true }` (or whatever the option is), otherwise this still won't work for ASGs.

Reply:
> 
> 
> Can you explain in a comment why the tag needs to be applied on `this`? That one surpised me.

Sure. It's a resource tag. So, all resources involved in the call must have the tag.

> For the other, don't forget to set ` { appliesToLaunchedInstances: true }` (or whatever the option is), otherwise this still won't work for ASGs.

Fortunately, it's `true` by default: https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.TagProps.html#applytolaunchedinstances